diff -up librsvg-2.26.0/tests/pdiff/Makefile.in.orig librsvg-2.26.0/tests/pdiff/Makefile.in
--- librsvg-2.26.0/tests/pdiff/Makefile.in.orig	2014-01-14 14:14:21.752647759 -0500
+++ librsvg-2.26.0/tests/pdiff/Makefile.in	2014-01-14 14:14:36.163544109 -0500
@@ -238,7 +238,7 @@ perceptualdiff_SOURCES = \
 	perceptualdiff.c
 
 INCLUDES = $(LIBRSVG_CFLAGS)
-LDADD = libpdiff.la $(LIBRSVG_LIBS)
+LDADD = libpdiff.la $(LIBRSVG_LIBS) -lm
 all: all-am
 
 .SUFFIXES:
